runtime/lua/vim/lsp.lua

```diff
@@ -0,0 +1,791 @@
+-- @param server_name [string] (optional)
+--
+-- @returns: success?, request_id, cancel_fn
+function lsp.buf_request(bufnr, method, params, callback)
+  if callback then
+    assert(type(callback) == 'function', "buf_request callback must be a function")
+  end
+  local client_request_ids = {}
+  for_each_buffer_client(bufnr, function(client, client_id)
+    local request_callback = callback
+    if not request_callback then
+      request_callback = client.default_server_callbacks[method]
+        or error(string.format("buf_request callback is empty and no default client was found for client %s", client.name))
+    end
+    local request_success, request_id = client.request(method, params, function(err, result)
+      -- TODO pass client here?
+      request_callback(err, result, client_id)
+    end)
+
+    -- This could only fail if the client shut down in the time since we looked
+    -- it up and we did the request, which should be rare.
+    if request_success then
+      client_request_ids[client_id] = request_id
+    end
+  end)
+
+  local function cancel_request()
+    for client_id, request_id in pairs(client_request_ids) do
+      local client = LSP_CLIENTS[client_id]
+      client.rpc.notify('$/cancelRequest', { id = request_id })
+    end
+  end
+
+  return client_request_ids, cancel_request
+end
+
+--- Send a request to a server, but don't wait for the response
+-- @param method [string]: Name of the request method
+-- @param params [string]: Arguments to send to the server
+-- @param cb [function|string] (optional): Either a function to call or a string to call in vim
+-- @param bufnr [number] (optional): The number of the buffer
+-- @param filetype [string] (optional): The filetype associated with the server
+-- @param server_name [string] (optional)
+--
+-- @returns: The table of request id
+function lsp.buf_request_sync(bufnr, method, params, timeout_ms)
+  local request_results = {}
+  local result_count = 0
+  local function callback(err, result, client_id)
+    _ = log.trace() and log.trace("callback", err, result, client_id)
+    request_results[client_id] = { error = err, result = result }
+    result_count = result_count + 1
+  end
+  local client_request_ids, cancel = lsp.buf_request(bufnr, method, params, callback)
+  _ = log.trace() and log.trace("client_request_ids", client_request_ids)
+
+  local expected_result_count = 0
+  for _ in pairs(client_request_ids) do
+    expected_result_count = expected_result_count + 1
+  end
+  _ = log.trace() and log.trace("expected_result_count", expected_result_count)
+  local timeout = (timeout_ms or 100) + vim.loop.now()
+  -- TODO is there a better way to sync this?
+  while result_count < expected_result_count do
+    _ = log.trace() and log.trace("results", result_count, request_results)
+    if vim.loop.now() >= timeout then
+      cancel()
+      return nil, "TIMEOUT"
+    end
+    -- TODO this really needs to be further looked at.
+    nvim_command "sleep 10m"
```

Comment:
we have `vim.fn.wait(condition)` but it only works for vimL callbacks. Though exposing the same in lua with lua cb would be straightforward.

Reply:
Then there is the question, if this is called from non-main coroutine, should it automagically yield instead (lua-client does this), or a separate `lsp.buf_request_yield` might be clearer/simpler.

Reply:
As long as the semantics and side effects are the same, I’d probably yield in a coroutine and wait otherwise. Fair point either way.